elastic/kibana
# [Fleet] Support preconfigured output secrets

## Summary

This PR adds support for outputs with secrets preconfigured in the `kibana.yml` config file.

As Kibana needs to compare the value of the secret to manage updates, a hash of the value is stored in the output's saved object. The implementation follows [option 2 in Infosec's recommendations](https://github.com/elastic/infosec/issues/14853#issuecomment-1788705606) with the Argon2id algorithm.

See [here](https://www.npmjs.com/package/argon2) for information about the `argon2` Node package and [here](https://github.com/ranisalt/node-argon2/wiki/Options) for the config options. Here, `argon2` was configured with the recommended `m=19456 (19 MiB), t=2, p=1 ` (for some reason, `timeCost` cannot be set to less than 2).

Closes #166360

### Testing

1. Ensure the [`outputSecretsStorage` experimental feature ](https://github.com/elastic/kibana/blob/fd4fdb01bcb011c0968a4ccab7ba739c3195016a/x-pack/plugins/fleet/common/experimental_features.ts#L26)is enabled.
2. Add the following to your kibana config:
   ```
   xpack.fleet.outputs:
     - id: my-logstash-output-with-a-secret
       name: preconfigured logstash output with a secret
       type:  logstash
       hosts: ["localhost:9999"]
       ssl:
         certificate: xxxxxxxxxx
       secrets:
         ssl:
           key: thisissecret
   ```
3. Verify the secret has been correctly created, e.g. by issuing a `GET .fleet-secrets/_search` request in Dev Tools: the secret should be listed there.
4. Change the preconfigured value and wait for kibana to restart: the secret should be updated with the new value.
